cpp/src/text/jaccard.cu

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <nvtext/detail/generate_ngrams.hpp>
+#include <nvtext/jaccard.hpp>
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+#include <thrust/execution_policy.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/transform.h>
+
+#include <cub/cub.cuh>
+
+namespace nvtext {
+namespace detail {
+namespace {
+
+/**
+ * @brief Retrieve the row data (span) for the given column/row-index
+ *
+ * @param d_input Input lists column
+ * @param idx Row index to retrieve
+ * @return A device-span of the row values
+ */
+__device__ auto get_row(cudf::column_device_view const& d_input, cudf::size_type idx)
+{
+  auto const offsets =
+    d_input.child(cudf::lists_column_view::offsets_column_index).data<cudf::size_type>();
+  auto const offset = offsets[idx];
+  auto const size   = offsets[idx + 1] - offset;
+  auto const begin =
+    d_input.child(cudf::lists_column_view::child_column_index).data<uint32_t>() + offset;
+  return cudf::device_span<uint32_t const>(begin, size);
+}
+
+/**
+ * @brief Count the unique values within each row of the input column
+ *
+ * This is called with a warp per row
+ */
+struct jaccard_unique_fn {
```

Comment:
Maybe more appropriate to call this something like `sorted_unique_fn` since it's not Jaccard-specific but it does assume that the data is presorted.